src/components/Main/state/serialization/encodeURIComponent.ts

```diff
@@ -0,0 +1,44 @@
+import type { AllParams } from '../../../../algorithms/types/Param.types'
+
+const serialize = (params: AllParams): string => {
+  return encodeURIComponent(JSON.stringify(params))
+}
+
+const deserialize = (queryString: string): AllParams => {
+  try {
+    const obj = JSON.parse(decodeURIComponent(queryString))
+
+    const simulationTimeRangeOriginal = obj.simulation.simulationTimeRange
+    const simulationTimeRangeConverted = {
+      tMin: new Date(simulationTimeRangeOriginal.tMin),
+      tMax: new Date(simulationTimeRangeOriginal.tMax),
+    }
+
+    // NOTE: Error in original parser. It should be: obj.containment
+    const containmentOriginal = obj.containment
+    const containmentConverted = {
+      ...containmentOriginal,
+      // NOTE: Error in original parser. It should be: containmentOriginal.reduction.map
+      reduction: containmentOriginal.map((c: { t: number; y: number }) => ({
+        y: c.y,
+        t: new Date(c.t),
+      })),
+    }
+
+    return {
+      ...obj,
+      containment: containmentConverted,
+      simulation: {
+        ...obj.simulation,
+        simulationTimeRange: simulationTimeRangeConverted,
+      },
+    }
+  } catch (error) {
+    throw new Error('encodeURIComponent: Error while parsing URL')
+  }
+}
```

Comment:
Do you think it would be possible to merge these serialize/deserialize functions and the serialize/deserialize functions in `jsUrl.ts` into a single deserialize function which accepts a `parse` function and a single serialize function which accepts a `stringify` function? I see that you have extra logic in the `jsUrl` serialize function to convert `Date` objects to numbers but I think that is the only difference? This probably isn't that big of deal as once I merged in my `io-ts` changes it will make serialization/deserialization much simpler.

Reply:
Yeah, we can massage it further afterwards. Currently there are 2 pairs of serializers because of the backward compatibility. Once we use only JSURL, this will be even simpler.

Dates are problematic, that’s why I would suggest to use timestamps starting from the DTO level.